flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,380 @@
+ * <p>On {@link #persist(SequenceNumber) persist}, accumulated changes are sent to the {@link
+ * StateChangeUploader} as an immutable {@link StateChangeUploader.UploadTask task}. An {@link
+ * FsStateChangelogWriter.UploadCompletionListener upload listener} is also registered. Upon
+ * notification it updates the Writer local state (for future persist calls) and completes the
+ * future returned to the original caller. The uploader notifies all listeners via a callback in a
+ * task.
+ *
+ * <p>If persist() is called for the same state change before the upload completion then the
+ * listener is added but not the upload task (which must already exist).
+ *
+ * <p>Invariants:
+ *
+ * <ol>
+ *   <li>Every change has at most one associated upload (retries are performed at a lower level)
+ *   <li>Every change is present in at most one collection: either uploaded OR notUploaded
+ *   <li>Changes BEING uploaded are NOT referenced locally - they will be added to uploaded upon
+ *       completion
+ *   <li>Failed and truncated changes are NOT stored - only their respective highest sequence
+ *       numbers
+ * </ol>
+ */
+@NotThreadSafe
+class FsStateChangelogWriter implements StateChangelogWriter<ChangelogStateHandleStreamImpl> {
+    private static final Logger LOG = LoggerFactory.getLogger(FsStateChangelogWriter.class);
+    private static final SequenceNumber INITIAL_SQN = SequenceNumber.of(0L);
+
+    private final UUID logId;
+    private final KeyGroupRange keyGroupRange;
+    private final StateChangeUploader uploader;
+    private final long preEmptivePersistThresholdInBytes;
+
+    /** Lock to synchronize handling of upload completion with new upload requests. */
+    // todo: replace with mailbox executor (after FLINK-23204)
+    private final Object lock = new Object();
+
+    /** A list of listener per upload (~ per checkpoint plus pre-emptive uploads). */
+    @GuardedBy("lock")
+    private final List<UploadCompletionListener> uploadCompletionListeners = new ArrayList<>();
+
+    private final FsStateChangelogCleaner cleaner;
+
+    /** Current {@link SequenceNumber}. */
+    private SequenceNumber activeSequenceNumber = INITIAL_SQN;
+
+    /**
+     * {@link SequenceNumber} before which changes will NOT be requested, exclusive. Increased after
+     * materialization.
+     */
+    @GuardedBy("lock")
+    private SequenceNumber lowestSequenceNumber = INITIAL_SQN;
+
+    /** Active changes, that will all use {@link #activeSequenceNumber}. */
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+
+    /** {@link #activeChangeSet} size in bytes. */
+    private long activeChangeSetSize;
+
+    /** Changes that are not yet uploaded (upload not requested). */
+    private final NavigableMap<SequenceNumber, StateChangeSet> notUploaded = new TreeMap<>();
+
+    /** Uploaded changes, ready for use in snapshots. */
+    @GuardedBy("lock")
+    private final NavigableMap<SequenceNumber, UploadResult> uploaded = new TreeMap<>();
```

Comment:
I'm lacking somewhere an explanation what is the relationship between activeChangeSet and notUploaded?

Reply:
The changes from `activeChangeSet` go to `notUploaded` in `rollover()` (I don't think this should be documented as such docs could easily become outdated, but I can add it if you prefer).

Reply:
I'll add a comment.
Renaming to something with "pending" I think would not improve clarity because there are multiple stages at which changes can be pending; and it's not clear what action is pending (sending to `Uploader` or moving to another collection).